firmware/powerboard2/code/app/ADC.c

```diff
@@ -0,0 +1,149 @@
+    [ADC_CHANNEL_ADC_BAT0_2]          = ADC_Channel_14,
+    [ADC_CHANNEL_ADC_BAT0_3]          = ADC_Channel_7,
+    [ADC_CHANNEL_ADC_BAT0_4]          = ADC_Channel_6,
+    [ADC_CHANNEL_ADC_BAT0_5]          = ADC_Channel_5,
+    [ADC_CHANNEL_ADC_BAT0_6]          = ADC_Channel_4,
+	[ADC_CHANNEL_ADC_BAT1_1]          = ADC_Channel_3,
+	[ADC_CHANNEL_ADC_BAT1_2]          = ADC_Channel_2,
+    [ADC_CHANNEL_ADC_BAT1_3]          = ADC_Channel_1,
+    [ADC_CHANNEL_ADC_BAT1_4]          = ADC_Channel_0,
+    [ADC_CHANNEL_ADC_BAT1_5]          = ADC_Channel_13,
+	[ADC_CHANNEL_ADC_BAT1_6]          = ADC_Channel_12,
+	[ADC_CHANNEL_ADC_EXT_PRESSURE]    = ADC_Channel_10,
+	[ADC_CHANNEL_ADC_WATER_SENSE]     = ADC_Channel_11,
+};
+
+/* PRIVATE FUNCTION DECLARATION */
+static void ADC_private_initGPIOs(void);
+static void ADC_private_initADC(void);
+static void ADC_private_initDMA(void);
+
+/* PRIVATE FUNCTION DEFINITION */
+static void ADC_private_initGPIOs(void)
+{
+    //Enable peripheral clock for GPIOA, GPIOB, and GPIOC
+	RCC_AHB1PeriphClockCmd(RCC_AHB1Periph_GPIOA, ENABLE);
+	RCC_AHB1PeriphClockCmd(RCC_AHB1Periph_GPIOC, ENABLE);
+
+    GPIO_InitTypeDef GPIO_InitStructure;
+    GPIO_StructInit(&GPIO_InitStructure);
+
+    GPIO_InitStructure.GPIO_Pin     = GPIO_Pin_0 | GPIO_Pin_1 | GPIO_Pin_2 | GPIO_Pin_3 | GPIO_Pin_4 | GPIO_Pin_5 | GPIO_Pin_6 | GPIO_Pin_7;
+    GPIO_InitStructure.GPIO_Mode    = GPIO_Mode_AN;
+    GPIO_InitStructure.GPIO_OType   = GPIO_OType_PP;
+    GPIO_InitStructure.GPIO_Speed   = GPIO_Speed_50MHz;
+    GPIO_InitStructure.GPIO_PuPd    = GPIO_PuPd_NOPULL;
+
+    GPIO_Init(GPIOA, &GPIO_InitStructure);
+
+    GPIO_InitStructure.GPIO_Pin     = GPIO_Pin_0 | GPIO_Pin_1 | GPIO_Pin_2 | GPIO_Pin_3 | GPIO_Pin_4 | GPIO_Pin_5;
+    
+    GPIO_Init(GPIOC, &GPIO_InitStructure);
+}
+
+static void ADC_private_initADC(void)
+{
+	RCC_APB2PeriphClockCmd(RCC_APB2Periph_ADC1, ENABLE);
+
+    ADC_InitTypeDef ADC_InitStructure;
+    ADC_StructInit(&ADC_InitStructure);
+
+    ADC_InitStructure.ADC_Resolution            = ADC_Resolution_12b;
+    ADC_InitStructure.ADC_ScanConvMode          = ENABLE;
+    ADC_InitStructure.ADC_ContinuousConvMode    = ENABLE;
+    ADC_InitStructure.ADC_ExternalTrigConvEdge  = ADC_ExternalTrigConvEdge_None;
+    ADC_InitStructure.ADC_DataAlign             = ADC_DataAlign_Right;
+    ADC_InitStructure.ADC_NbrOfConversion       = (uint8_t)ADC_CHANNEL_COUNT;
+
+    // Writes the settings above into the the ADC config registers
+    ADC_Init(ADC1, &ADC_InitStructure);
+
+    // Choose which HW ADC channels to sample and the sampling rate.
+    for(ADC_channel_E channel = (ADC_channel_E)0U; channel < ADC_CHANNEL_COUNT; channel++)
+    {
+    	ADC_RegularChannelConfig(ADC1, ADC_softwareChannelToADCChannelMapping[channel], channel + 1, ADC_SampleTime_28Cycles);
+    }
+
+    // Enable the ADC
+    ADC_Cmd(ADC1, ENABLE);
+
+    // DMA will not stop after first ADC conversion stops
+    ADC_DMARequestAfterLastTransferCmd(ADC1, ENABLE);
```

Comment:
I think this is only needed for Single ADC mode. I think we can remove this and it will still work.

Reply:
I was wrong about this. Turns out you need it. Can you add it back please?